### PR TITLE
[Proof] Introduce AccumulatorRangeProof

### PR DESCRIPTION
**Pull request by wqfish**
## Motivation

I wanted to make things more consistent. `AccumulatorProof` is a generic struct that proves one leaf exists in a Merkle accumulator and `SignedTransactionWithProof` is an object that contains an instance of this proof. Similarly, `AccumulatorRangeProof` is a generic struct that proves a list of leaves exist in a Merkle accumulator, and `TransactionListWithProof` will contain an instance of this proof.

This also means that the proof does not depend on the data -- it doesn't matter if data is a list of transaction or a list of events. This way the tests for the proof verification will be easier to write.

This change also removes some of the unnecessary hashes in the proof to make it a bit smaller.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y.

## Test Plan

CI.

## Related PRs

None.

### CODE REVIEW

File: types/src/proof/definition.rs [hunk 1]
```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
...
+            "Proof has more than {} ({}) right siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.right_siblings.len(),
+        );
+        ensure!(
+            !leaf_hashes.is_empty(),
+            "leaf_hashes is empty while first_leaf_index indicated non-empty list.",
+        );
+
+        let mut left_sibling_iter = self.left_siblings.iter().rev().peekable();
```
Comment (msmouse) [inline types/src/proof/definition.rs]: unrelated: we `reverse()`ed in `get_siblings()` and `rev()`s here -- seems a waste. Shall we change the proof format (separately)?

File: types/src/proof/definition.rs [hunk 2]
```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
...
+            first_leaf_index.expect("first_leaf_index should not be None."),
+        );
+        let mut current_hashes = leaf_hashes.to_vec();
+        let mut parent_hashes = vec![];
+
+        // Keep reducing the list of hashes by combining all the children pairs, until there is
+        // only one hash left.
+        while current_hashes.len() > 1
+            || left_sibling_iter.peek().is_some()
+            || right_sibling_iter.peek().is_some()
```
Comment (msmouse) [inline types/src/proof/definition.rs]: Do you feel it's better to calculate the root level and loop by num of levels up (and assert siblings are fully consumed afterwards)? Probably it's stricter and more straightforward that way.

File: types/src/proof/definition.rs [hunk 3]
```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
...
+        {
+            let mut leaf_iter = current_hashes.iter();
```
Comment (msmouse) [inline types/src/proof/definition.rs]: Name `leaf_iter` is probably a little confusing?

File: types/src/proof/definition.rs [hunk 4]
```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
...
+            // If the first leaf is a right child, it needs to be combined with a sibling on the
```
Comment (msmouse) [inline types/src/proof/definition.rs]: "first posistion"

File: types/src/proof/definition.rs [hunk 5]
```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
...
+                })?;
+                let right_hash = *leaf_iter.next().expect("The first leaf must exist.");
+                parent_hashes.push(MerkleTreeInternalNode::<H>::new(left_hash, right_hash).hash());
+            }
+
+            // Similarly, if the last leaf is a left child, it needs to be combined with a sibling
+            // on the right. Everything else should form an even number of leaves.
+            for chunk in leaf_iter.as_slice().chunks(2) {
+                let left_hash = chunk[0];
+                let right_hash = if chunk.len() == 2 {
```
Comment (msmouse) [inline types/src/proof/definition.rs]: (I'm not sure, but:) Will it be more clear if we use `chunks_exact()` and `remainder()`?

Comment (wqfish) [reply types/src/proof/definition.rs]: Probably.

```diff
             // Similarly, if the last leaf is a left child, it needs to be combined with a sibling
             // on the right. Everything else should form an even number of leaves.
-            for chunk in leaf_iter.as_slice().chunks(2) {
+            let mut leaf_iter = leaf_iter.as_slice().chunks_exact(2);
+            while let Some(chunk) = leaf_iter.next() {
                 let left_hash = chunk[0];
-                let right_hash = if chunk.len() == 2 {
-                    chunk[1]
-                } else {
-                    *right_sibling_iter.next().ok_or_else(|| {
-                        format_err!("Last child is a left child, but missing sibling on the right.")
-                    })?
-                };
+                let right_hash = chunk[1];
+                parent_hashes.push(MerkleTreeInternalNode::<H>::new(left_hash, right_hash).hash());
+            }
+
+            let remainder = leaf_iter.remainder();
+            assert!(remainder.len() <= 1);
+            if !remainder.is_empty() {
+                let left_hash = remainder[0];
+                let right_hash = *right_sibling_iter.next().ok_or_else(|| {
+                    format_err!("Last child is a left child, but missing sibling on the right.")
+                })?;
                 parent_hashes.push(MerkleTreeInternalNode::<H>::new(left_hash, right_hash).hash());
             }

```

File: types/src/proof/definition.rs (hunk 2 above)
Comment (wqfish) [reply types/src/proof/definition.rs]: I was following `AccumulatorProof::verify` which does it in a similar way...

File: types/src/proof/definition.rs (hunk 1 above)
Comment (wqfish) [reply types/src/proof/definition.rs]: I don't remember why we set up things in this way. @lightmark ?

File: storage/accumulator/src/lib.rs [hunk 6]
```diff
@@ -353,6 +355,61 @@ where
...
+        first_leaf_index: Option<u64>,
+        num_leaves: LeafCount,
+    ) -> Result<AccumulatorRangeProof<H>> {
+        if first_leaf_index.is_none() {
+            ensure!(
+                num_leaves == 0,
+                "num_leaves is not zero while first_leaf_index is None.",
+            );
+            return Ok(AccumulatorRangeProof::new(vec![], vec![]));
+        }
```
[the remark this answers is not in the document]
Comment (wqfish) [reply storage/accumulator/src/lib.rs]: If you ask for zero items, the proof doesn't need to contain anything... I don't think this part is important, just for completeness.

**Pull request closed without merging**